src/common/btree.c

```diff
@@ -4150,8 +4488,20 @@ btr_class_init(umem_off_t root_off, struct btr_root *root,
 	if (tc->tc_feats & BTR_FEAT_SKIP_LEAF_REBAL)
 		*tree_feats |= BTR_FEAT_SKIP_LEAF_REBAL;
 
-	/** Only check btree managed bits */
-	if ((*tree_feats & tc->tc_feats) != (*tree_feats & BTR_FEAT_MASK)) {
+	if ((*tree_feats & (BTR_FEAT_UINT_KEY | BTR_FEAT_EMBED_FIRST)) ==
+	    (BTR_FEAT_UINT_KEY | BTR_FEAT_EMBED_FIRST)) {
+		/** Integer keys are incompatible with this optimization so
+		 * disable it.  The reason is we want to embed the user
+		 * allocated offset in the tree root.  The key is normally
+		 * stored in this record but with integer keys, it's stored
+		 * in the btr_record. Therefore, we can't save an allocation
+		 * with integer keys making this optimization useless.
```

Comment:
Some confused. As my understand, this optimization saves the root node allocation for the first record, even if it is BTR_FEAT_UINT_KEY tree. If without such optimization, the BTR_FEAT_UINT_KEY tree also needs twice allocation for the first record: one is for the record itself, the other is for the root node, right?

Reply:
It would save root node allocation.  That root node, for dynamic trees (our default) would have one entry.  In the case of hashed keys, the key is stored in the value.  With integer keys, it's stored in the record (where the hash is normally stored).  In order to support integer keys, we would need to add 8 bytes to the record allocation to store that key because we have no place otherwise to store it.  I decided it wasn't worth the effort.   Theoretically, we could do it, yes.